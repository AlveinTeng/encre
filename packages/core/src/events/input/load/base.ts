import { Context } from './context';

/**
 * Interface that defines the events for loading readable contexts.
 */
export interface Loader {
  load(): Promise<Context[]>;
}

/**
 * Abstract class that provides an abstract load() event from the
 * Loader interface.
 */
export abstract class BaseLoader implements Loader {
  /**
   * Abstract method that loads the readable contexts.
   * @returns A Promise that resolves with an array of Context instances.
   */
  abstract load(): Promise<Context[]>;
}
