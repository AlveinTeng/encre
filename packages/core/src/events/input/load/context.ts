export type ContextMetadata = Record<string, unknown>;

/**
 * Interface for interacting with a readable context.
 */
export interface ContextInput<ContextMetadata = Record<string, unknown>> {
  pageContent: string;

  metadata?: ContextMetadata;
}

/**
 * Class for constructing a readable context with content in one page.
 */
export class Context<ContextMetadata = Record<string, unknown>>
  implements ContextInput<ContextMetadata>
{
  pageContent: string;

  metadata: ContextMetadata;

  constructor(fields: ContextInput<ContextMetadata>) {
    this.pageContent = fields.pageContent
      ? fields.pageContent.toString()
      : this.pageContent;
    this.metadata = fields.metadata ?? ({} as ContextMetadata);
  }
}
