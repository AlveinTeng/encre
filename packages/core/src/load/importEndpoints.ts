// This file is generated by `scripts/create-entrypoints.js`. DO NOT edit manually.

export const optionalImportEndpoints = [
  'events/input/load/docs/buffer',
  'events/input/load/docs/pdf'
] as string[];
