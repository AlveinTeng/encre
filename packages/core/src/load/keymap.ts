import camelcase from 'camelcase';
import decamelize from 'decamelize';
// import snakeCase from 'decamelize';

export type SerializedFields = {
  [key: string]: unknown;
};

export type SecretFields = {
  [key: string]: string;
};

export type SerializedKeyAlias = {
  [key: string]: string;
};

export function keyToJson(key: string, map?: SerializedKeyAlias): string {
  return map?.[key] || decamelize(key);
}

export function keyFromJson(key: string, map?: SerializedKeyAlias): string {
  return map?.[key] || camelcase(key);
}

export function mapKeys(
  fields: SerializedFields,
  mapper: typeof keyFromJson,
  map?: SerializedKeyAlias
): SerializedFields {
  const mapToKeys: SerializedFields = {};

  for (const key in fields) {
    // console.log(`Map Keys: key: ${key}, fields: ${JSON.stringify(fields)}`);
    if (Object.hasOwn(fields, key)) {
      mapToKeys[mapper(key, map)] = fields[key];
    }
  }

  return mapToKeys;
}
