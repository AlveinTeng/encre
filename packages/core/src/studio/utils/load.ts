import { FileLoader, loadFromFile } from '../../utils/load.js';
import { parseFile } from '../../utils/parse.js';
import { BaseGraph } from '../graph.js';
import { GuardrailRegistration } from '../registration/guardrails.js';
import { NodeRegistration } from '../registration/nodes.js';

export const loadGraphFromFile: FileLoader<BaseGraph> = async (
  file: string,
  path: string,
  values: Record<string, unknown> = {},
  registry?: {
    nodes?: NodeRegistration;
    guardrails?: GuardrailRegistration;
  }
): Promise<BaseGraph> => {
  const serialized = parseFile(file, path);
  return BaseGraph.deserialize(serialized, values, registry);
};

export const loadGraph = async (
  uri: string,
  values: Record<string, unknown> = {},
  registry?: {
    nodes?: NodeRegistration;
    guardrails?: GuardrailRegistration;
  }
): Promise<BaseGraph> => {
  return loadFromFile(uri, loadGraphFromFile, values, registry);
};
