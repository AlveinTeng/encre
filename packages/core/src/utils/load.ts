import * as yaml from 'yaml';

export type FileLoader<T> = (
  file: string,
  path: string,
  values: Record<string, unknown>,
  ...args
) => Promise<T>;

export const loadFromFile = async <T>(
  uri: string,
  loader: FileLoader<T>,
  values: Record<string, unknown> = {},
  ...args
): Promise<T> => {
  try {
    const fs = await import('node:fs/promises');
    return loader(
      await fs.readFile(uri, { encoding: 'utf-8' }),
      uri,
      values,
      ...args
    );
  } catch (e) {
    console.error(e);
    throw new Error(`Could not load file at ${uri}`);
  }
};

export const loadFileContents = (contents: string, format: string) => {
  switch (format) {
    case '.json':
      return JSON.parse(contents);
    case '.encre':
    case '.yml':
    case '.yaml':
      return yaml.parse(contents);
    default:
      throw new Error(`Unsupported filetype ${format}`);
  }
};
