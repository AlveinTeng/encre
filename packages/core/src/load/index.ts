import { getEnvironmentVariables } from '../utils/environment';
import { optionalImportEndpoints } from './importEndpoints.js';
/*eslint import/namespace: ['error', { allowComputed: true }]*/
import * as importMap from './importMap.js';
import { OptionalImportMap, SecretMap } from './importType.js';
import {
  SerializedFields,
  SerializedKeyAlias,
  keyFromJson,
  mapKeys,
} from './keymap.js';
import {
  SerializableType,
  Serializable,
  SerializedConstructor,
  SerializedErrnoRecord,
  SerializedEventRecord,
  SerializedInputRecord,
  SerializedNotImplemented,
  SerializedRecord,
  SerializedSecret,
  SerializedSecretRecord,
  getUniqueName,
} from './serializable.js';

type ImportModule = (typeof importMap)[keyof typeof importMap];

type OptionalImportModule = OptionalImportMap[keyof OptionalImportMap];

function combineAliasesAndInvert(constructor: typeof Serializable) {
  const aliases: SerializedKeyAlias = {};

  for (
    let current = constructor;
    current && current.prototype;
    current = Object.getPrototypeOf(current)
  ) {
    Object.assign(aliases, Reflect.get(current.prototype, '_aliases'));
  }

  return Object.entries(aliases).reduce(
    (accumulator, [key, value]) => {
      accumulator[value] = key;

      return accumulator;
    },
    {} as Record<string, string>
  );
}

async function build(
  optionalImportsMap: OptionalImportMap,
  name: string,
  namespace: string[],
  pathStr: string,
  serializedStr: string
) {
  let module: ImportModule | OptionalImportModule | null = null;
  const endpoint: string = namespace.join('/');

  if (
    optionalImportEndpoints.includes(endpoint) ||
    endpoint in optionalImportsMap
  ) {
    if (endpoint in optionalImportsMap) {
      module =
        await optionalImportsMap[endpoint as keyof typeof optionalImportsMap];
    } else {
      throw new Error(
        `[MODULE] Missing key "${endpoint}" for ${pathStr} in load(optionalImportsMap={})`
      );
    }
  } else {
    if (namespace.length === 0) {
      throw new Error(
        `[MODULE] Invalid namespace: ${pathStr} -> ${serializedStr}`
      );
    }

    let importMapKey: string;
    do {
      importMapKey = namespace.join('_');
      if (importMapKey in importMap) {
        break;
      } else {
        namespace.pop();
      }
    } while (namespace.length > 0);

    if (importMapKey in importMap) {
      module = importMap[importMapKey as keyof typeof importMap];
    }
  }

  if (module === null || typeof module !== 'object') {
    throw new Error(
      `[MODULE] Invalid namespace: ${pathStr} -> ${serializedStr}`
    );
  }

  const builder =
    module[name as keyof typeof module] ??
    Object.values(module).find(
      (v) =>
        typeof v === 'function' &&
        getUniqueName(v as typeof Serializable) === name
    );

  if (typeof builder !== 'function') {
    throw new Error(
      `[MODULE] Invalid identifier: ${pathStr} -> ${serializedStr}`
    );
  }

  return builder;
}

async function reviveSecret(
  this: {
    optionalImportsMap: OptionalImportMap;
    secretsMap: SecretMap;
    path?: string[];
  },
  value: unknown
) {
  const { secretsMap, path = ['$'] } = this;
  const pathStr = path.join('.');

  const serialized = value as SerializedSecret;
  const [key] = serialized._id;

  if (key in secretsMap) {
    return secretsMap[key as keyof SecretMap];
  } else {
    const secretValueInEnv: string | undefined = getEnvironmentVariables(key);
    if (secretValueInEnv) {
      return secretValueInEnv;
    } else {
      throw new Error(
        `[MODULE] Missing key "${key}" for ${pathStr} in load(secretsMap={})`
      );
    }
  }
}

async function reviveConstructor(
  this: {
    optionalImportsMap: OptionalImportMap;
    secretsMap: SecretMap;
    path?: string[];
  },
  value: unknown
) {
  const { optionalImportsMap, path = ['$'] } = this;
  const pathStr = path.join('.');

  const serialized = value as SerializedConstructor;
  const serializedStr: string = JSON.stringify(serialized);
  const [name, ...namespaceReverse] = serialized._id.slice().reverse();
  const namespace: string[] = namespaceReverse.reverse();

  const builder = await build(
    optionalImportsMap,
    name,
    namespace,
    pathStr,
    serializedStr
  );

  const kwargs = await reviver.call(
    { ...this, path: [...path, 'kwargs'] },
    serialized._kwargs
  );

  if (serialized._type === 'constructor') {
    const instance = new (builder as any)(
      mapKeys(
        kwargs as SerializedFields,
        keyFromJson,
        combineAliasesAndInvert(builder as typeof Serializable)
      )
    );

    Object.defineProperty(instance.constructor, 'name', { value: name });

    return instance;
  } else {
    throw new Error(
      `[MODULE] Invalid identifier: ${pathStr} -> ${serializedStr}`
    );
  }
}

async function reviveSecretRecord(
  this: {
    optionalImportsMap: OptionalImportMap;
    secretsMap: SecretMap;
    path?: string[];
  },
  value: unknown,
) {
  const { secretsMap, path = ['$'] } = this;
  const pathStr = path.join('.');

  const serialized = value as SerializedSecretRecord;
  const [key] = serialized._id;

  if (key in secretsMap) {
    return secretsMap[key as keyof SecretMap];
  } else {
    const secretValueInEnv: string | undefined = getEnvironmentVariables(key);
    if (secretValueInEnv) {
      return secretValueInEnv;
    } else {
      throw new Error(
        `[RECORD] Missing key "${key}" for ${pathStr} in load(secretsMap={})`
      );
    }
  }
}

async function reviveRecord(
  this: {
    optionalImportsMap: OptionalImportMap;
    secretsMap: SecretMap;
    path?: string[];
  },
  value: unknown
) {
  const { optionalImportsMap, path = ['$'] } = this;
  const pathStr = path.join('.');

  const serialized = value as SerializedRecord;
  const serializedStr: string = JSON.stringify(serialized);
  const [name, ...namespaceReverse] = serialized._id.slice().reverse();
  const namespace: string[] = namespaceReverse.reverse();

  const builder = await build(
    optionalImportsMap,
    name,
    namespace,
    pathStr,
    serializedStr
  );

  if (serialized._type === 'event_record') {
    const serializedEvent = serialized as SerializedEventRecord;
    const kwargs = await reviver.call(
      { ...this, path: [...path, 'kwargs'] },
      serializedEvent._kwargs
    );
    const instance = new (builder as any)(
      mapKeys(
        kwargs as SerializedFields,
        keyFromJson,
        combineAliasesAndInvert(builder as typeof Serializable)
      )
    );

    Object.defineProperty(instance.constructor, 'name', { value: name });

    return instance;
  } else {
    throw new Error(
      `[RECORD] Invalid identifier type: ${serialized._type}: ${pathStr} -> ${serializedStr}`
    );
  }
}

async function reviver(
  this: {
    optionalImportsMap: OptionalImportMap;
    secretsMap: SecretMap;
    path?: string[];
  },
  value: unknown
) {
  const { path = ['$'] } = this;
  const pathStr = path.join('.');

  if (
    typeof value === 'object' &&
    value !== null &&
    !Array.isArray(value) &&
    '_grp' in value &&
    '_type' in value &&
    '_id' in value &&
    '_recordId' in value &&
    value._grp === 2
  ) {
    if (value._type === 'errno_record') {
      const serialized = value as SerializedErrnoRecord;
      const serializedStr: string = JSON.stringify(serialized);
      throw new Error(
        `[RECORD] Trying to load an record that produce errno serialization: ${pathStr} -> ${serializedStr}`
      );
    } else if (value._type === 'secret_record') {
      return await reviveSecretRecord.call(this, value);
    } else {
      return await reviveRecord.call(this, value);
    }
  } else if (
    typeof value === 'object' &&
    value !== null &&
    !Array.isArray(value) &&
    '_grp' in value &&
    '_type' in value &&
    '_id' in value &&
    value._grp === 1
  ) {
    if (value._type === 'not_implemented') {
      const serialized = value as SerializedNotImplemented;
      const serializedStr: string = JSON.stringify(serialized);
      throw new Error(
        `[MODULE] Trying to load an object that doesn't implement serialization: ${pathStr} -> ${serializedStr}`
      );
    } else if (value._type === 'secret') {
      return await reviveSecret.call(this, value);
    } else {
      return await reviveConstructor.call(this, value);
    }
  } else if (typeof value === 'object' && value !== null) {
    if (Array.isArray(value)) {
      return Promise.all(
        value.map((v, i) =>
          reviver.call({ ...this, path: [...path, `${i}`] }, v)
        )
      );
    } else {
      return Object.fromEntries(
        await Promise.all(
          Object.entries(value).map(async ([k, v]) => [
            k,
            await reviver.call({ ...this, path: [...path, k] }, v),
          ])
        )
      );
    }
  }

  return value;
}

export async function load<T>(
  text: string,
  secretsMap: SecretMap = {},
  optionalImportsMap: OptionalImportMap = {}
): Promise<T> {
  const json = JSON.parse(text);
  return reviver.call({ secretsMap, optionalImportsMap }, json) as Promise<T>;
}
