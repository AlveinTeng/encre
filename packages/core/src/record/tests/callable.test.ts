import { expect, test } from "@jest/globals";
import { stringify } from "yaml";
import { load } from "../../load";
import { OptionalImportMap, SecretMap } from "../../load/importType.js";
import { SecretFields } from "../../load/keymap";
import { SerializedConstructor } from "../../load/serializable";
import {
  Callable,
  CallableConfigFields,
  CallableConfig,
  CallableBindArgs,
} from "../callable";

test("test custom callable", async () => {
  type TestInput = {
    input: string | boolean;
  };

  interface TestParams extends CallableConfig {
    hello: string;
    bye: string;
    testApi: string;
  }

  interface SerializableTestParams extends TestParams {
    serializableField: boolean;
  }

  class Test<
      CallOptions extends SerializableTestParams = SerializableTestParams,
    >
    extends Callable<TestInput, string, CallOptions>
    implements TestParams
  {
    _namespace: string[] = ["tests"];

    _isCallable = true;

    _isSerializable = true;

    get _secrets(): SecretFields | undefined {
      return {
        testApi: "TEST_API",
      };
    }

    get _aliases(): SecretFields | undefined {
      return {
        testApi: "test_api",
      };
    }

    hello: string;

    bye: string;

    testApi: string;

    constructor(fields?: Partial<TestInput> & SerializableTestParams) {
      super(fields ?? {});
      this.hello = fields?.hello ?? "default-hello";
      this.bye = fields?.bye ?? "default-bye";
      this.testApi = fields?.testApi ?? "default-api";
    }

    async invoke(
      input: TestInput,
      options?: Partial<CallOptions> | undefined
    ): Promise<string> {
      this.hello = options?.hello ?? this.hello;
      this.bye = options?.bye ?? this.bye;

      return `${this.hello}-${String(input.input)}-${this.bye}-${String(
        this._kwargs.serializableField
      )}`;
    }
  }

  const test = new Test({
    hello: "hello",
    bye: "bye",
    serializableField: true,
    testApi: "this-is-api",
  });
  const argumentsBefore = test._kwargs;
  const serializedStr: string = JSON.stringify(test, null, 2);

  expect(test._kwargs).toEqual(argumentsBefore);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  expect(await test.invoke({ input: "input" })).toBe("hello-input-bye-true");
  expect(await test.invoke({ input: true })).toBe("hello-true-bye-true");

  const test2 = await load<Test>(
    serializedStr,
    { TEST_API: "this-is-a-key" } as SecretMap,
    { tests: { Test } } as OptionalImportMap
  );

  expect(test2).toBeInstanceOf(Test);
  expect(JSON.stringify(test2, null, 2)).toBe(serializedStr);

  expect(await test2.invoke({ input: "input" })).toBe("hello-input-bye-true");
  expect(await test2.invoke({ input: true })).toBe("hello-true-bye-true");

  expect(
    await test2.invoke(
      { input: "input" },
      { hello: "new-hello", bye: "new-bye" }
    )
  ).toBe("new-hello-input-new-bye-true");
  expect(
    await test2.invoke({ input: true }, { hello: "new-hello", bye: "new-bye" })
  ).toBe("new-hello-true-new-bye-true");

  const record = await test.toRecord("output");
  const recordStr = JSON.stringify(record, null, 2);
  expect(JSON.parse(recordStr)).toMatchSnapshot({
    _recordId: expect.any(String),
    _metadata: {
      _secrets: {
        _recordId: expect.any(String),
      },
      _inputs: {
        _recordId: expect.any(String),
      },
      _outputs: {
        _recordId: expect.any(String),
      },
    },
  });

  const test3 = await load<Test>(
    recordStr,
    { TEST_API: "this-is-a-key" } as SecretMap,
    { tests: { Test } } as OptionalImportMap
  );

  expect(test3).toBeInstanceOf(Test);
  expect(JSON.stringify(test3, null, 2)).toBe(serializedStr);
});

test("test CallableBind", async () => {
  type TestInput = {
    input: string | boolean;
  };

  interface TestParams extends CallableConfig {
    hello: string;
    bye: string;
    testApi: string;
  }

  interface SerializableTestParams extends TestParams {
    serializableField: boolean;
  }

  class Test<
      CallOptions extends SerializableTestParams = SerializableTestParams,
    >
    extends Callable<TestInput, string, CallOptions>
    implements TestParams
  {
    _namespace: string[] = ["tests"];

    _isCallable = true;

    _isSerializable = true;

    get _secrets(): SecretFields | undefined {
      return {
        testApi: "TEST_API",
      };
    }

    get _aliases(): SecretFields | undefined {
      return {
        testApi: "test_api",
      };
    }

    hello: string;

    bye: string;

    testApi: string;

    constructor(fields?: Partial<TestInput> & SerializableTestParams) {
      super(fields ?? {});
      this.hello = fields?.hello ?? "default-hello";
      this.bye = fields?.bye ?? "default-bye";
      this.testApi = fields?.testApi ?? "default-api";
    }

    async invoke(
      input: TestInput,
      options?: Partial<CallOptions> | undefined
    ): Promise<string> {
      this.hello = options?.hello ?? this.hello;
      this.bye = options?.bye ?? this.bye;

      return `${this.hello}-${String(input.input)}-${this.bye}-${String(
        this._kwargs.serializableField
      )}`;
    }
  }

  class TestBind<
      CallOptions extends SerializableTestParams = SerializableTestParams,
    >
    extends Callable<TestInput, string, CallOptions>
    implements TestParams
  {
    _namespace: string[] = ["tests"];

    _isCallable = true;

    _isSerializable = true;

    get _secrets(): SecretFields | undefined {
      return {
        testApi: "TEST_API",
      };
    }

    get _aliases(): SecretFields | undefined {
      return {
        testApi: "test_api",
      };
    }

    hello: string;

    bye: string;

    testApi: string;

    constructor(fields?: Partial<TestInput> & SerializableTestParams) {
      super(fields ?? {});
      this.hello = fields?.hello ?? "default-hello";
      this.bye = fields?.bye ?? "default-bye";
      this.testApi = fields?.testApi ?? "default-api";
    }

    async invoke(
      input: TestInput,
      options?: Partial<CallOptions> | undefined
    ): Promise<string> {
      return "this-is-a-bind";
    }
  }

  const test = new Test({
    hello: "hello",
    bye: "bye",
    serializableField: true,
    testApi: "this-is-api",
  });

  const testbind = new TestBind({
    hello: "bind-hello",
    bye: "bind-bye",
    serializableField: true,
    testApi: "bind-this-is-api",
  });

  const tBind = testbind.bind(test);
  const newTest = test.bind(testbind);

  expect(await tBind.invoke({ input: "input" })).toBe("this-is-a-bind");
  expect(await newTest.invoke({ input: "input" })).toBe(
    "bind-hello-input-bind-bye-true"
  );

  const serializedStr: string = JSON.stringify(newTest, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  expect(
    stringify(JSON.parse(JSON.stringify(tBind, null, 2)))
  ).toMatchSnapshot();

  const record = await newTest.toRecord("output");
  const recordStr = JSON.stringify(record, null, 2);
  expect(JSON.parse(recordStr)).toMatchSnapshot({
    _recordId: expect.any(String),
    _metadata: {
      _inputs: {
        _recordId: expect.any(String),
        _kwargs: {
          bound: {
            _recordId: expect.any(String),
            _metadata: {
              _secrets: {
                _recordId: expect.any(String),
              },
              _inputs: {
                _recordId: expect.any(String),
              },
              _outputs: {
                _recordId: expect.any(String),
              },
            },
          },
        },
      },
      _outputs: {
        _recordId: expect.any(String),
      },
    },
  });

  const newBindObj = JSON.parse(serializedStr) as SerializedConstructor;
  const newBindArgs = newBindObj._kwargs as CallableBindArgs<
    TestInput,
    string,
    SerializableTestParams
  >;

  const newTest2 = await load<Test>(
    JSON.stringify(newBindArgs.bound, null, 2),
    { TEST_API: 'this-is-a-key' } as SecretMap,
    { tests: { Test } } as OptionalImportMap
  );
  expect(newTest2).toBeInstanceOf(Test);

  const newTest3 = await load<TestBind>(
    JSON.stringify(newBindArgs.kwargs, null, 2),
    { TEST_API: 'this-is-a-key' } as SecretMap,
    { tests: { TestBind } } as OptionalImportMap
  );
  expect(newTest3).toBeInstanceOf(TestBind);
});

test("test CallableEach", async () => {
  type TestInput = {
    input: string | boolean;
  };

  interface TestParams extends CallableConfig {
    hello: string;
    bye: string;
    testApi: string;
  }

  interface SerializableTestParams extends TestParams {
    serializableField: boolean;
  }

  class Test<
      CallOptions extends SerializableTestParams = SerializableTestParams,
    >
    extends Callable<TestInput, string, CallOptions>
    implements TestParams
  {
    _namespace: string[] = ["tests"];

    _isCallable = true;

    _isSerializable = true;

    get _secrets(): SecretFields | undefined {
      return {
        testApi: "TEST_API",
      };
    }

    get _aliases(): SecretFields | undefined {
      return {
        testApi: "test_api",
      };
    }

    hello: string;

    bye: string;

    testApi: string;

    constructor(fields?: Partial<TestInput> & SerializableTestParams) {
      super(fields ?? {});
      this.hello = fields?.hello ?? "default-hello";
      this.bye = fields?.bye ?? "default-bye";
      this.testApi = fields?.testApi ?? "default-api";
    }

    async invoke(
      input: TestInput,
      options?: Partial<CallOptions> | undefined
    ): Promise<string> {
      this.hello = options?.hello ?? this.hello;
      this.bye = options?.bye ?? this.bye;

      return `${this.hello}-${String(input.input)}-${this.bye}-${String(
        this._kwargs.serializableField
      )}`;
    }
  }

  const test = new Test({
    hello: "hello",
    bye: "bye",
    serializableField: true,
    testApi: "this-is-api",
  });

  const testMap = test.map();
  const serializedStr: string = JSON.stringify(testMap, null, 2);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();
});
