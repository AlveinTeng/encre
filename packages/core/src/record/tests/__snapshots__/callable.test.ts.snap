// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test CallableBind 1`] = `
"_grp: 1
_type: constructor
_id:
  - record
  - callable
  - CallableBind
_kwargs:
  bound:
    _grp: 1
    _type: constructor
    _id:
      - tests
      - Test
    _kwargs:
      hello: bind-hello
      bye: bind-bye
      serializable_field: true
      test_api:
        _grp: 1
        _type: secret
        _id:
          - TEST_API
  kwargs:
    _grp: 1
    _type: constructor
    _id:
      - tests
      - TestBind
    _kwargs:
      hello: bind-hello
      bye: bind-bye
      serializable_field: true
      test_api:
        _grp: 1
        _type: secret
        _id:
          - TEST_API
  config: {}
"
`;

exports[`test CallableBind 2`] = `
"_grp: 1
_type: constructor
_id:
  - record
  - callable
  - CallableBind
_kwargs:
  bound:
    _grp: 1
    _type: constructor
    _id:
      - tests
      - TestBind
    _kwargs:
      hello: bind-hello
      bye: bind-bye
      serializable_field: true
      test_api:
        _grp: 1
        _type: secret
        _id:
          - TEST_API
  kwargs:
    _grp: 1
    _type: constructor
    _id:
      - tests
      - Test
    _kwargs:
      hello: bind-hello
      bye: bind-bye
      serializable_field: true
      test_api:
        _grp: 1
        _type: secret
        _id:
          - TEST_API
  config: {}
"
`;

exports[`test CallableBind 3`] = `
{
  "_grp": 2,
  "_id": [
    "record",
    "callable",
    "CallableBind",
  ],
  "_kwargs": {
    "bound": {
      "_grp": 1,
      "_id": [
        "tests",
        "Test",
      ],
      "_kwargs": {
        "bye": "bind-bye",
        "hello": "bind-hello",
        "serializable_field": true,
        "test_api": {
          "_grp": 1,
          "_id": [
            "TEST_API",
          ],
          "_type": "secret",
        },
      },
      "_type": "constructor",
    },
    "config": {},
    "kwargs": {
      "_grp": 1,
      "_id": [
        "tests",
        "TestBind",
      ],
      "_kwargs": {
        "bye": "bind-bye",
        "hello": "bind-hello",
        "serializable_field": true,
        "test_api": {
          "_grp": 1,
          "_id": [
            "TEST_API",
          ],
          "_type": "secret",
        },
      },
      "_type": "constructor",
    },
  },
  "_metadata": {
    "_inputs": {
      "_grp": 2,
      "_id": [
        "record",
        "callable",
        "CallableBind",
      ],
      "_kwargs": {
        "bound": {
          "_grp": 2,
          "_id": [
            "tests",
            "Test",
          ],
          "_kwargs": {
            "bye": "bind-bye",
            "hello": "bind-hello",
            "serializable_field": true,
            "test_api": {
              "_grp": 1,
              "_id": [
                "TEST_API",
              ],
              "_type": "secret",
            },
          },
          "_metadata": {
            "_inputs": {
              "_grp": 2,
              "_id": [
                "tests",
                "Test",
              ],
              "_kwargs": {
                "bye": "string",
                "hello": "string",
                "serializable_field": "boolean",
              },
              "_recordId": Any<String>,
              "_type": "input_record",
            },
            "_outputs": {
              "_grp": 2,
              "_id": [
                "tests",
                "Test",
              ],
              "_kwargs": {
                "outputs": "string",
              },
              "_recordId": Any<String>,
              "_type": "output_record",
            },
            "_recordType": "event",
            "_secrets": {
              "_grp": 2,
              "_id": [
                "TEST_API",
              ],
              "_recordId": Any<String>,
              "_type": "secret_record",
            },
          },
          "_recordId": Any<String>,
          "_type": "event_record",
        },
        "config": {},
        "kwargs": {
          "_grp": 1,
          "_id": [
            "tests",
            "TestBind",
          ],
          "_kwargs": {
            "bye": "bind-bye",
            "hello": "bind-hello",
            "serializable_field": true,
            "test_api": {
              "_grp": 1,
              "_id": [
                "TEST_API",
              ],
              "_type": "secret",
            },
          },
          "_type": "constructor",
        },
      },
      "_recordId": Any<String>,
      "_type": "input_record",
    },
    "_outputs": {
      "_grp": 2,
      "_id": [
        "record",
        "callable",
        "CallableBind",
      ],
      "_kwargs": {
        "outputs": "string",
      },
      "_recordId": Any<String>,
      "_type": "output_record",
    },
    "_recordType": "event",
    "_secrets": {},
  },
  "_recordId": Any<String>,
  "_type": "event_record",
}
`;

exports[`test CallableEach 1`] = `
"_grp: 1
_type: constructor
_id:
  - record
  - callable
  - CallableEach
_kwargs:
  bound:
    _grp: 1
    _type: constructor
    _id:
      - tests
      - Test
    _kwargs:
      hello: hello
      bye: bye
      serializable_field: true
      test_api:
        _grp: 1
        _type: secret
        _id:
          - TEST_API
"
`;

exports[`test custom callable 1`] = `
"_grp: 1
_type: constructor
_id:
  - tests
  - Test
_kwargs:
  hello: hello
  bye: bye
  serializable_field: true
  test_api:
    _grp: 1
    _type: secret
    _id:
      - TEST_API
"
`;

exports[`test custom callable 2`] = `
{
  "_grp": 2,
  "_id": [
    "tests",
    "Test",
  ],
  "_kwargs": {
    "bye": "bye",
    "hello": "hello",
    "serializable_field": true,
    "test_api": {
      "_grp": 1,
      "_id": [
        "TEST_API",
      ],
      "_type": "secret",
    },
  },
  "_metadata": {
    "_inputs": {
      "_grp": 2,
      "_id": [
        "tests",
        "Test",
      ],
      "_kwargs": {
        "bye": "string",
        "hello": "string",
        "serializable_field": "boolean",
      },
      "_recordId": Any<String>,
      "_type": "input_record",
    },
    "_outputs": {
      "_grp": 2,
      "_id": [
        "tests",
        "Test",
      ],
      "_kwargs": {
        "outputs": "string",
      },
      "_recordId": Any<String>,
      "_type": "output_record",
    },
    "_recordType": "event",
    "_secrets": {
      "_grp": 2,
      "_id": [
        "TEST_API",
      ],
      "_recordId": Any<String>,
      "_type": "secret_record",
    },
  },
  "_recordId": Any<String>,
  "_type": "event_record",
}
`;
